trainer/trainer.py

```diff
@@ -1,346 +1,606 @@
+        """Reads:
+        ```yaml
+        stagename:
+          - dataset1 frac
+          - dataset2 frac
+          - until dataset3 epochs
+        ```
+        """
+        return {
+            stage_name: self._load_stage(ymldata, stage_name, datasets, int(ymldata['seed']))
+            for stage_name in stages_order
+        }
+
+    def _load_stage(self, ymldata:dict, stage_name:str, available_datasets:Dict[str,Dataset], seed:int) -> Stage:
+        datasets: List[Tuple[Dataset, float]] = []
+
+        for line in ymldata[stage_name][:-1]:
+            dataset_name, weight = line.split()
+            datasets.append((available_datasets[dataset_name], float(weight)))
+
+        _, dataset_name, max_epochs = ymldata[stage_name][-1].split()
+        assert dataset_name in available_datasets, f"until clause of stage '{stage_name}' refers to unknown dataset '{dataset_name}'"
+        assert dataset_name in {dataset.name for dataset, weight in datasets if weight > 0.0}, f"until clause of stage '{stage_name}' watches dataset '{dataset_name}' but that dataset is not read during this stage"
+        return Stage(
+            name=stage_name,
+            datasets=datasets,
+            until_dataset=dataset_name,
+            until_epoch=int(max_epochs) if max_epochs != 'inf' else None
+        )
+
+    def _load_modifiers(self, ymldata:dict) -> List[Modifier]:
+        """Reads
+        ```yml
+        uppercase: 0.05
+        titlecase: 0.05
+        ```
+        """
+        return [
+            Modifier(name, float(ymldata[name]))
+            for name in MODIFIERS.keys()
+            if name in ymldata
+        ]
+
+
+class CurriculumV2Loader(CurriculumV1Loader):
+    """Slightly different curriculum format that can have multiple files per
+    dataset, and puts the modifiers in their own section."""
+
+    def _load_datasets(self, ymldata:dict, basepath:str) -> Dict[str,Dataset]:
+        """Reads
+        ```yml
+        datasets:
+          clean:
+            - a.gz
+            - b.gz
+        ```
+        """
+        return {
+            name: Dataset(name, [
+                os.path.join(basepath, filepath)
+                for filepath in files
+            ])
+            for name, files in ymldata['datasets']
+        }
+
+    def _load_modifiers(self, ymldata:dict) -> List[Modifier]:
+        """Reads
+        ```yml
+        modifiers:
+          - uppercase 0.05
+          - titlecase 0.05
+        ```
+        """
+        return [
+            self._load_modifier(modifier_line)
+            for modifier_line in ymldata.get('modifiers', [])
+        ]
+
+    def _load_modifier(self, line:str) -> Modifier:
+        name, frequency = line.split()
+        return Modifier(name, float(frequency))
+
+
+class CurriculumLoader:
+    """Reads curriculum yaml files. Wrapper that decides which reader to use
+    based on a version number that may be in there.
+    """
+
+    IMPLEMENTATIONS={
```

Comment:
We haven't released it so maybe we can make this a possibility, although the way i imagine it, i don't see a reason why we should have multiple unmerged files per stage...?

Reply:
Collapsed it all into a version 1. I've left the IMPLEMENTATIONS switch there for future versions. Modifiers will need a rework some time in the future when there are more options, but I don't have the best ideas for that right now.